# Project 1: Trading with Momentum
## Instructions
Each problem consists of a function to implement and instructions on how to implement the function.  The parts of the function that need to be implemented are marked with a `# TODO` comment. After implementing the function, run the cell to test it against the unit tests we've provided. For each problem, we provide one or more unit tests from our `project_tests` package. These unit tests won't tell you if your answer is correct, but will warn you of any major errors. Your code will be checked for the correct solution when you submit it to Udacity.

## Packages
When you implement the functions, you'll only need to you use the packages you've used in the classroom, like [Pandas](https://pandas.pydata.org/) and [Numpy](http://www.numpy.org/). These packages will be imported for you. We recommend you don't add any import statements, otherwise the grader might not be able to run your code.

The other packages that we're importing are `helper`, `project_helper`, and `project_tests`. These are custom packages built to help you solve the problems.  The `helper` and `project_helper` module contains utility functions and graph functions. The `project_tests` contains the unit tests for all the problems.

### Install Packages

In [1]:
import sys
!{sys.executable} -m pip install -r requirements.txt

    100% |████████████████████████████████| 880kB 9.9MB/s eta 0:00:01    77% |█████████████████████████       | 686kB 12.3MB/s eta 0:00:01
    100% |████████████████████████████████| 17.0MB 2.1MB/s eta 0:00:01  4% |█▍                              | 737kB 16.3MB/s eta 0:00:01    9% |███                             | 1.5MB 17.5MB/s eta 0:00:01    29% |█████████▋                      | 5.1MB 19.5MB/s eta 0:00:01    40% |█████████████                   | 6.9MB 17.2MB/s eta 0:00:01    50% |████████████████▏               | 8.6MB 19.2MB/s eta 0:00:01    55% |█████████████████▊              | 9.4MB 18.8MB/s eta 0:00:01    66% |█████████████████████▏          | 11.2MB 21.3MB/s eta 0:00:01    72% |███████████████████████         | 12.2MB 20.5MB/s eta 0:00:01    77% |████████████████████████▊       | 13.1MB 22.3MB/s eta 0:00:01    88% |████████████████████████████▏   | 15.0MB 18.1MB/s eta 0:00:01
    100% |████████████████████████████████| 26.2MB 1.6MB/s eta 0:00:01 0% |                         

    100% |████████████████████████████████| 81kB 10.9MB/s ta 0:00:01
  Running setup.py bdist_wheel for cvxpy ... done
  Stored in directory: /root/.cache/pip/wheels/2b/60/0b/0c2596528665e21d698d6f84a3406c52044c7b4ca6ac737cf3
  Running setup.py bdist_wheel for plotly ... done
  Stored in directory: /root/.cache/pip/wheels/98/54/81/dd92d5b0858fac680cd7bdb8800eb26c001dd9f5dc8b1bc0ba
  Running setup.py bdist_wheel for scs ... done
  Stored in directory: /root/.cache/pip/wheels/df/d0/79/37ea880586da03c620ca9ecd5e42adbd86bc6ea84363965c5f
Successfully built cvxpy plotly scs
tensorflow 1.3.0 requires tensorflow-tensorboard<0.2.0,>=0.1.0, which is not installed.
moviepy 0.2.3.2 has requirement tqdm==4.11.2, but you'll have tqdm 4.19.5 which is incompatible.
  Found existing installation: numpy 1.12.1
    Uninstalling numpy-1.12.1:
      Successfully uninstalled numpy-1.12.1
  Found existing installation: scipy 1.2.1
    Uninstalling scipy-1.2.1:
      Successfully uninstalled scipy-1.2.1
  Fou

### Load Packages

In [2]:
import pandas as pd
import numpy as np
import helper
import project_helper
import project_tests

## Market Data
### Load Data
The data we use for most of the projects is end of day data. This contains data for many stocks, but we'll be looking at stocks in the S&P 500. We also made things a little easier to run by narrowing down our range of time period instead of using all of the data.

In [3]:
df = pd.read_csv('../../data/project_1/eod-quotemedia.csv', parse_dates=['date'], index_col=False)

close = df.reset_index().pivot(index='date', columns='ticker', values='adj_close')

print('Loaded Data')

Loaded Data


### View Data
Run the cell below to see what the data looks like for `close`.

In [5]:
project_helper.print_dataframe(close)

### Stock Example
Let's see what a single stock looks like from the closing prices. For this example and future display examples in this project, we'll use Apple's stock (AAPL). If we tried to graph all the stocks, it would be too much information.

In [6]:
apple_ticker = 'AAPL'
project_helper.plot_stock(close[apple_ticker], '{} Stock'.format(apple_ticker))

## Resample Adjusted Prices

The trading signal you'll develop in this project does not need to be based on daily prices, for instance, you can use month-end prices to perform trading once a month. To do this, you must first resample the daily adjusted closing prices into monthly buckets, and select the last observation of each month.

Implement the `resample_prices` to resample `close_prices` at the sampling frequency of `freq`.

In [14]:
def resample_prices(close_prices, freq='M'):
    """
    Resample close prices for each ticker at specified frequency.
    
    Parameters
    ----------
    close_prices : DataFrame
        Close prices for each ticker and date
    freq : str
        What frequency to sample at
        For valid freq choices, see http://pandas.pydata.org/pandas-docs/stable/timeseries.html#offset-aliases
    
    Returns
    -------
    prices_resampled : DataFrame
        Resampled prices for each ticker and date
    """
    # TODO: Implement Function
    
    return pd.DataFrame(close_prices.resample(freq).last())

project_tests.test_resample_prices(resample_prices)

Tests Passed


In [15]:
resample_prices(close)

ticker               A         AAL          AAP         AAPL        ABBV  \
date                                                                       
2013-07-31 30.77861719 18.63139292  81.73270857  58.73000866 38.52144972   
2013-08-31 32.09288410 15.55986096  79.33492514  63.64994327 36.09056668   
2013-09-30 35.34697923 18.25587647  81.98212977  62.28266407 37.88620154   
2013-10-31 35.00902763 21.15409315  98.34285959  68.28583759 41.39637606   
2013-11-30 36.94707663 22.60801580 100.15741326  73.07037475 41.39637606   
2013-12-31 39.53485221 24.31228275 109.80605374  73.72082947 45.12162270   
2014-01-31 40.19849023 32.30404302 113.90344263  65.78133976 42.40046983   
2014-02-28 39.35511692 35.55851889 126.35434591  69.56208595 43.84740848   
2014-03-31 38.65691442 35.24077420 125.56160936  70.95004943 44.26943225   
2014-04-30 37.44602782 33.76759430 120.39025770  78.00222579 45.26058952   
2014-05-31 39.45552968 38.66856535 123.24889355  84.14255587 47.21597213   
2014-06-30 39.89336706 41.36458087 133.98342494  86.47010113 49.04968649   
2014-07-31 38.95576181 37.50350151 120.26928991  88.95449982 45.83414944   
2014-08-31 39.69890079 37.53342708 135.47301238  95.84930753 48.40870808   
2014-09-30 39.66634256 34.25030202 129.45327437  94.21285594 50.58044463   
2014-10-31 38.48289605 40.01590351 146.00501306 100.99244110 55.99653498   
2014-11-30 41.16610236 46.96425145 146.12423326 111.69411705 61.06145951   
2014-12-31 39.43238724 51.89970750 158.30385803 103.66431211 57.74366923   
2015-01-31 36.46881439 47.49650651 158.02557400 110.03180655 53.66393519   
2015-02-28 40.75585532 46.44954689 153.98051686 121.11705716 53.79731697   
2015-03-31 40.21618088 51.18177630 148.83252640 117.31741727 52.05446174   
2015-04-30 40.04195916 46.91986067 142.18084892 117.99626112 57.97557525   
2015-05-31 39.86773744 41.17209272 152.34230539 123.34294830 59.70605561   
2015-06-30 37.43482962 38.80593634 158.43730330 118.74646370 60.24402878   
2015-07-31 39.73448090 38.96627137 173.27743492 114.84110860 63.23267196   
2015-08-31 35.23221004 37.96773731 174.31186769 107.23801710 56.36839104   
2015-09-30 33.40767610 37.82163263 188.57875989 104.56562594 49.14283218   
2015-10-31 36.74552431 45.01972341 197.43409130 113.64795179 54.29511396   
2015-11-30 40.69644669 40.27495546 161.91326754 112.99053695 53.01865452   
2015-12-31 40.79862571 41.33893271 149.81538015 100.53578968 54.01246937   
2016-01-31 36.73925516 38.05914962 151.34827289  92.97124993 50.61292076   
2016-02-29 36.44651209 40.13373983 147.75493343  92.84754723 50.34556653   
2016-03-31 38.88603766 40.14352854 159.66025522 104.65874623 52.65956346   
2016-04-30 40.04407095 33.95705938 155.43823026  90.01478000 56.77932921   
2016-05-31 44.90768367 31.32651866 153.18781130  96.47778143 58.57578996   
2016-06-30 43.52297301 27.79234545 161.00763909  92.36206594 57.62636511   
2016-07-31 47.20221441 34.85087472 169.20594924 100.68044865 62.18997907   
2016-08-31 46.09353634 35.74268695 156.77400383 103.05832580 60.18990877   
2016-09-30 46.31429082 36.04792753 148.60671775 109.80908324 59.22273863   
2016-10-31 42.85227545 39.97666915 139.59783411 110.28503592 52.86517692   
2016-11-30 43.25552156 45.84312511 169.13581100 107.90875084 57.62285329   
2016-12-31 44.93909238 46.08991196 168.59615688 113.08352807 59.34774792   
2017-01-31 48.30261971 43.68127231 163.73127251 118.48287110 58.52291275   
2017-02-28 50.60086565 45.86299598 156.12490616 134.32908516 59.22200826   
2017-03-31 52.27967020 41.84652135 147.85939394 140.86952606 62.40145631   
2017-04-30 54.43532901 42.16309078 141.75593049 140.85972030 63.77814835   
2017-05-31 59.66626253 47.99551598 133.26892494 150.40575387 63.85552554   
2017-06-30 58.77942143 49.88916265 116.33305213 141.80044954 70.13275003   

ticker              ABC         ABT          ACN         ADBE         ADI  \
date                                                                        
2013-07-31  53.87744989 32.99081455  66.22844

### View Data
Let's apply this function to `close` and view the results.

In [16]:
monthly_close = resample_prices(close)
project_helper.plot_resampled_prices(
    monthly_close.loc[:, apple_ticker],
    close.loc[:, apple_ticker],
    '{} Stock - Close Vs Monthly Close'.format(apple_ticker))

## Compute Log Returns

Compute log returns ($R_t$) from prices ($P_t$) as your primary momentum indicator:

$$R_t = log_e(P_t) - log_e(P_{t-1})$$

Implement the `compute_log_returns` function below, such that it accepts a dataframe (like one returned by `resample_prices`), and produces a similar dataframe of log returns. Use Numpy's [log function](https://docs.scipy.org/doc/numpy/reference/generated/numpy.log.html) to help you calculate the log returns.

In [17]:
np.log(monthly_close)

ticker              A        AAL        AAP       AAPL       ABBV        ABC  \
date                                                                           
2013-07-31 3.42682020 2.92484795 4.40345427 4.07295082 3.65121522 3.98671202   
2013-08-31 3.46863433 2.74469458 4.37367845 4.15339843 3.58603152 3.97061867   
2013-09-30 3.56521294 2.90448703 4.40650129 4.13168312 3.63458697 4.04148377   
2013-10-31 3.55560596 3.05183342 4.58845994 4.22370239 3.72319334 4.10842325   
2013-11-30 3.60948653 3.11830452 4.60674308 4.29142301 3.72319334 4.18839927   
2013-12-31 3.67718262 3.19098169 4.69871566 4.30028538 3.80936157 4.18527516   
2014-01-31 3.69382944 3.47519239 4.73535110 4.18633621 3.74715944 4.14033195   
2014-02-28 3.67262600 3.57117976 4.83909023 4.24221968 3.78071562 4.15311601   
2014-03-31 3.65472566 3.56220377 4.83279655 4.26197610 3.79029442 4.11923987   
2014-04-30 3.62290064 3.51950159 4.79073861 4.35673736 3.81243667 4.11296930   
2014-05-31 3.67517421 3.65502701 4.81420584 4.43251245 3.85473223 4.23221422   
2014-06-30 3.68621007 3.72242498 4.89771610 4.45979870 3.89283379 4.22508309   
2014-07-31 3.66242669 3.62443430 4.78973331 4.48812500 3.82502943 4.28192797   
2014-08-31 3.68132350 3.62523192 4.90877245 4.56277724 3.87967972 4.29121203   
2014-09-30 3.68050303 3.53369538 4.86332000 4.54555665 3.92356503 4.29004841   
2014-10-31 3.65021388 3.68927696 4.98364096 4.61504567 4.02528981 4.38981764   
2014-11-30 3.71761516 3.84938670 4.98445717 4.71576404 4.11188089 4.45701347   
2014-12-31 3.67458749 3.94931315 5.06451634 4.64115791 4.05601372 4.44719054   
2015-01-31 3.59645750 3.86065616 5.06275688 4.70076947 3.98274118 4.50000774   
2015-02-28 3.70759952 3.83836671 5.03682608 4.79675749 3.98522360 4.58090157   
2015-03-31 3.69426942 3.93538354 5.00282169 4.76488323 3.95229051 4.68180491   
2015-04-30 3.68992788 3.84844105 4.95709983 4.77065294 4.06002181 4.68733196   
2015-05-31 3.68556741 3.71776067 5.02613000 4.81496867 4.08943345 4.67451783   
2015-06-30 3.62260154 3.65857323 5.06535895 4.77699066 4.09840346 4.61767293   
2015-07-31 3.68221935 3.66269644 5.15489398 4.74354951 4.14682113 4.61210924   
2015-08-31 3.56196072 3.63673678 5.16084604 4.67505082 4.03190856 4.55938932   
2015-09-30 3.50878570 3.63288123 5.23951574 4.64981487 3.89473100 4.50759083   
2015-10-31 3.60401643 3.80710069 5.28540481 4.73310553 3.99443424 4.52346584   
2015-11-30 3.70614078 3.69572982 5.08706081 4.72730407 3.97064382 4.54879834   
2015-12-31 3.70864840 3.72180474 5.00940374 4.61051378 3.98921493 4.59892002   
2016-01-31 3.60384581 3.63914152 5.01958362 4.53229030 3.92420689 4.45223027   
2016-02-29 3.59584576 3.69221737 4.99555505 4.53095887 3.91891056 4.42276657   
2016-03-31 3.66063526 3.69246125 5.07304815 4.65070502 3.96384786 4.42195811   
2016-04-30 3.68998062 3.52509677 5.04624842 4.49997388 4.03917234 4.40506287   
2016-05-31 3.80460891 3.44446498 5.03166469 4.56931274 4.07032147 4.28297732   
2016-06-30 3.77328891 3.32476064 5.08145181 4.52571635 4.05398019 4.33924622   
2016-07-31 3.85444081 3.55107824 5.13111661 4.61195163 4.13019388 4.41063997   
2016-08-31 3.83067273 3.57634569 5.05480530 4.63529510 4.09750471 4.43513337   
2016-09-30 3.83545057 3.58484937 5.00130334 4.69874325 4.08130557 4.36129955   
2016-10-31 3.75775875 3.68829601 4.93876568 4.70306825 3.96774484 4.22262639   
2016-11-30 3.76712489 3.82522524 5.13070201 4.68128597 4.05391925 4.33069381   
2016-12-31 3.80530807 3.83059410 5.12750625 4.72812673 4.08341417 4.33325496   
2017-01-31 3.87748580 3.77691946 5.09822650 4.77476840 4.06941835 4.44323454   
2017-02-28 3.92396868 3.82565860 5.05065637 4.90029265 4.08129323 4.49459450   
2017-03-31 3.95660758 3.73400867 4.99626178 4.94783411 4.13358861 4.46114879   
2017-04-30 3.99701337 3.74154521 4.95410678 4.94776450 4.15541063 4.38547506   
2017-05-31 4.08876674 3.87110759 4.89236908 5.01333667 4.15662312 4.50163326   
2017-06-30 4.07379182 3.90980380 4.75645722 4.95442078 4.25038988 4.53126505   


In [20]:
def compute_log_returns(prices):
    """
    Compute log returns for each ticker.
    
    Parameters
    ----------
    prices : DataFrame
        Prices for each ticker and date
    
    Returns
    -------
    log_returns : DataFrame
        Log returns for each ticker and date
    """
    # TODO: Implement Function
    
    return (np.log(prices) - np.log(prices.shift(1)))

project_tests.test_compute_log_returns(compute_log_returns)

Tests Passed


### View Data
Using the same data returned from `resample_prices`, we'll generate the log returns.

In [21]:
monthly_close_returns = compute_log_returns(monthly_close)
project_helper.plot_returns(
    monthly_close_returns.loc[:, apple_ticker],
    'Log Returns of {} Stock (Monthly)'.format(apple_ticker))

## Shift Returns
Implement the `shift_returns` function to shift the log returns to the previous or future returns in the time series. For example, the parameter `shift_n` is 2 and `returns` is the following:

```
                           Returns
               A         B         C         D
2013-07-08     0.015     0.082     0.096     0.020     ...
2013-07-09     0.037     0.095     0.027     0.063     ...
2013-07-10     0.094     0.001     0.093     0.019     ...
2013-07-11     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...
```

the output of the `shift_returns` function would be:
```
                        Shift Returns
               A         B         C         D
2013-07-08     NaN       NaN       NaN       NaN       ...
2013-07-09     NaN       NaN       NaN       NaN       ...
2013-07-10     0.015     0.082     0.096     0.020     ...
2013-07-11     0.037     0.095     0.027     0.063     ...
...            ...       ...       ...       ...
```
Using the same `returns` data as above, the `shift_returns` function should generate the following with `shift_n` as -2:
```
                        Shift Returns
               A         B         C         D
2013-07-08     0.094     0.001     0.093     0.019     ...
2013-07-09     0.092     0.057     0.069     0.087     ...
...            ...       ...       ...       ...       ...
...            ...       ...       ...       ...       ...
...            NaN       NaN       NaN       NaN       ...
...            NaN       NaN       NaN       NaN       ...
```
_Note: The "..." represents data points we're not showing._

In [23]:
def shift_returns(returns, shift_n):
    """
    Generate shifted returns
    
    Parameters
    ----------
    returns : DataFrame
        Returns for each ticker and date
    shift_n : int
        Number of periods to move, can be positive or negative
    
    Returns
    -------
    shifted_returns : DataFrame
        Shifted returns for each ticker and date
    """
    # TODO: Implement Function
    
    return returns.shift(shift_n)

project_tests.test_shift_returns(shift_returns)

Tests Passed


### View Data
Let's get the previous month's and next month's returns.

In [41]:
prev_returns = shift_returns(monthly_close_returns, 1)
lookahead_returns = shift_returns(monthly_close_returns, -1)

project_helper.plot_shifted_returns(
    prev_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Previous Returns of {} Stock'.format(apple_ticker))
project_helper.plot_shifted_returns(
    lookahead_returns.loc[:, apple_ticker],
    monthly_close_returns.loc[:, apple_ticker],
    'Lookahead Returns of {} Stock'.format(apple_ticker))

## Generate Trading Signal

A trading signal is a sequence of trading actions, or results that can be used to take trading actions. A common form is to produce a "long" and "short" portfolio of stocks on each date (e.g. end of each month, or whatever frequency you desire to trade at). This signal can be interpreted as rebalancing your portfolio on each of those dates, entering long ("buy") and short ("sell") positions as indicated.

Here's a strategy that we will try:
> For each month-end observation period, rank the stocks by _previous_ returns, from the highest to the lowest. Select the top performing stocks for the long portfolio, and the bottom performing stocks for the short portfolio.

Implement the `get_top_n` function to get the top performing stock for each month. Get the top performing stocks from `prev_returns` by assigning them a value of 1. For all other stocks, give them a value of 0. For example, using the following `prev_returns`:

```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0.015     0.082     0.096     0.020     0.075     0.043     0.074
2013-07-09     0.037     0.095     0.027     0.063     0.024     0.086     0.025
...            ...       ...       ...       ...       ...       ...       ...
```

The function `get_top_n` with `top_n` set to 3 should return the following:
```
                                     Previous Returns
               A         B         C         D         E         F         G
2013-07-08     0         1         1         0         1         0         0
2013-07-09     0         1         0         1         0         1         0
...            ...       ...       ...       ...       ...       ...       ...
```
*Note: You may have to use Panda's [`DataFrame.iterrows`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.DataFrame.iterrows.html) with [`Series.nlargest`](https://pandas.pydata.org/pandas-docs/version/0.21/generated/pandas.Series.nlargest.html) in order to implement the function. This is one of those cases where creating a vecorization solution is too difficult.*

In [74]:
def get_top_n(prev_returns, top_n):
    """
    Select the top performing stocks
    
    Parameters
    ----------
    prev_returns : DataFrame
        Previous shifted returns for each ticker and date
    top_n : int
        The number of top performing stocks to get
    
    Returns
    -------
    top_stocks : DataFrame
        Top stocks for each ticker and date marked with a 1
    """
    # TODO: Implement Function
    top_stocks = pd.DataFrame(0, index=prev_returns.index, columns=prev_returns.columns)
    for index, series in prev_returns.iterrows():
        top_stocks.loc[index][series.nlargest(top_n).index] = 1
    return top_stocks

project_tests.test_get_top_n(get_top_n)

Tests Passed


### View Data
We want to get the best performing and worst performing stocks. To get the best performing stocks, we'll use the `get_top_n` function. To get the worst performing stocks, we'll also use the `get_top_n` function. However, we pass in `-1*prev_returns` instead of just `prev_returns`. Multiplying by negative one will flip all the positive returns to negative and negative returns to positive. Thus, it will return the worst performing stocks.

In [75]:
top_bottom_n = 50
df_long = get_top_n(prev_returns, top_bottom_n)
df_short = get_top_n(-1*prev_returns, top_bottom_n)
project_helper.print_top(df_long, 'Longed Stocks')
project_helper.print_top(df_short, 'Shorted Stocks')

10 Most Longed Stocks:
INCY, AMD, AVGO, NFX, SWKS, NFLX, ILMN, UAL, NVDA, MU
10 Most Shorted Stocks:
RRC, FCX, CHK, MRO, GPS, WYNN, DVN, FTI, SPLS, TRIP


## Projected Returns
It's now time to check if your trading signal has the potential to become profitable!

We'll start by computing the net returns this portfolio would return. For simplicity, we'll assume every stock gets an equal dollar amount of investment. This makes it easier to compute a portfolio's returns as the simple arithmetic average of the individual stock returns.

Implement the `portfolio_returns` function to compute the expected portfolio returns. Using `df_long` to indicate which stocks to long and `df_short` to indicate which stocks to short, calculate the returns using `lookahead_returns`. To help with calculation, we've provided you with `n_stocks` as the number of stocks we're investing in a single period.

In [89]:
def portfolio_returns(df_long, df_short, lookahead_returns, n_stocks):
    """
    Compute expected returns for the portfolio, assuming equal investment in each long/short stock.
    
    Parameters
    ----------
    df_long : DataFrame
        Top stocks for each ticker and date marked with a 1
    df_short : DataFrame
        Bottom stocks for each ticker and date marked with a 1
    lookahead_returns : DataFrame
        Lookahead returns for each ticker and date
    n_stocks: int
        The number number of stocks chosen for each month
    
    Returns
    -------
    portfolio_returns : DataFrame
        Expected portfolio returns for each ticker and date
    """
    # TODO: Implement Function
          
    return ((lookahead_returns * df_long) - (lookahead_returns * df_short))/n_stocks

project_tests.test_portfolio_returns(portfolio_returns)

Tests Passed


### View Data
Time to see how the portfolio did.

In [90]:
expected_portfolio_returns = portfolio_returns(df_long, df_short, lookahead_returns, 2*top_bottom_n)
project_helper.plot_returns(expected_portfolio_returns.T.sum(), 'Portfolio Returns')

## Statistical Tests
### Annualized Rate of Return

In [91]:
expected_portfolio_returns_by_date = expected_portfolio_returns.T.sum().dropna()
portfolio_ret_mean = expected_portfolio_returns_by_date.mean()
portfolio_ret_ste = expected_portfolio_returns_by_date.sem()
portfolio_ret_annual_rate = (np.exp(portfolio_ret_mean * 12) - 1) * 100

print("""
Mean:                       {:.6f}
Standard Error:             {:.6f}
Annualized Rate of Return:  {:.2f}%
""".format(portfolio_ret_mean, portfolio_ret_ste, portfolio_ret_annual_rate))


Mean:                       0.003253
Standard Error:             0.002203
Annualized Rate of Return:  3.98%



The annualized rate of return allows you to compare the rate of return from this strategy to other quoted rates of return, which are usually quoted on an annual basis. 

### T-Test
Our null hypothesis ($H_0$) is that the actual mean return from the signal is zero. We'll perform a one-sample, one-sided t-test on the observed mean return, to see if we can reject $H_0$.

We'll need to first compute the t-statistic, and then find its corresponding p-value. The p-value will indicate the probability of observing a t-statistic equally or more extreme than the one we observed if the null hypothesis were true. A small p-value means that the chance of observing the t-statistic we observed under the null hypothesis is small, and thus casts doubt on the null hypothesis. It's good practice to set a desired level of significance or alpha ($\alpha$) _before_ computing the p-value, and then reject the null hypothesis if $p < \alpha$.

For this project, we'll use $\alpha = 0.05$, since it's a common value to use.

Implement the `analyze_alpha` function to perform a t-test on the sample of portfolio returns. We've imported the `scipy.stats` module for you to perform the t-test.

Note: [`scipy.stats.ttest_1samp`](https://docs.scipy.org/doc/scipy-1.0.0/reference/generated/scipy.stats.ttest_1samp.html) performs a two-sided test, so divide the p-value by 2 to get 1-sided p-value

In [95]:
from scipy import stats

def analyze_alpha(expected_portfolio_returns_by_date):
    """
    Perform a t-test with the null hypothesis being that the expected mean return is zero.
    
    Parameters
    ----------
    expected_portfolio_returns_by_date : Pandas Series
        Expected portfolio returns for each date
    
    Returns
    -------
    t_value
        T-statistic from t-test
    p_value
        Corresponding p-value
    """
    # TODO: Implement Function
    null_hypothesis = 0.0
    t_value, p_value = stats.ttest_1samp(expected_portfolio_returns_by_date.values, null_hypothesis)

    return (t_value, p_value/2)

project_tests.test_analyze_alpha(analyze_alpha)

Tests Passed


### View Data
Let's see what values we get with our portfolio. After you run this, make sure to answer the question below.

In [96]:
t_value, p_value = analyze_alpha(expected_portfolio_returns_by_date)
print("""
Alpha analysis:
 t-value:        {:.3f}
 p-value:        {:.6f}
""".format(t_value, p_value))


Alpha analysis:
 t-value:        1.476
 p-value:        0.073359



### Question: What p-value did you observe? And what does that indicate about your signal?

p-value of 0.073359 - this is above 0.05 (5%), meaning that the model does not produce statistically significant returns compared to our null hypothesis that the mean should be 0.0.

## Submission
Now that you're done with the project, it's time to submit it. Click the submit button in the bottom right. One of our reviewers will give you feedback on your project with a pass or not passed grade. You can continue to the next section while you wait for feedback.